addons/mail/models/mail_message.py

```diff
@@ -906,7 +905,6 @@ def _generate_model_record_ids(msg_val, msg_ids):
                 mid for mid, message in message_values.items()
                 if (message.get('model') == model and
                     message.get('res_id') in mids.ids and
-                    message.get('message_type') != 'user_notification' and
```

Comment:
I don't think it's correct, if you can read a record, should you be able to read notifications sent to other users about this document? I don't now if some use case exists where it is a problem but it doesn't look right. 
Plus there is nothing about this in the commit message. Maybe the message unlink should be done in sudo if it is a problem.

Reply:
Indeed. To be honest the code for mail messages ACL/rules is so hard to review, it's important to explain in details the behavior change, and to cover the changes with updated tests to indicate that they are indeed the expected behavior.

Using sudo exclusively during deletion of user_notifications would have the advantage of never allowing such deletions outside of a full unlink() of the record. i.e. never a single message.

I'm not opposed to changing the code that checks access when needed, but we need to confirm that it is indeed implementing a desired set of rules.